db/memtable.cc

```diff
@@ -842,6 +846,66 @@ bool MemTable::Get(const LookupKey& key, std::string* value, Status* s,
   return found_final_value;
 }
 
+bool MemTable::MultiGet(const ReadOptions& read_options, MultiGetRange* range,
+                        ReadCallback* callback, bool* is_blob) {
+  // The sequence number is updated synchronously in version_set.h
+  if (IsEmpty()) {
+    // Avoiding recording stats for speed.
+    return false;
+  }
+  std::bitset<MultiGetContext::MAX_BATCH_SIZE> keys;
+  int idx = 0;
+  for (auto iter = range->begin(); iter != range->end(); ++iter) {
+    Slice user_key = iter->lkey->user_key();
+    size_t ts_sz =
+        GetInternalKeyComparator().user_comparator()->timestamp_size();
+    bool may_contain = true;
+    if (bloom_filter_) {
+      // when both memtable_whole_key_filtering and prefix_extractor_ are set,
+      // only do whole key filtering for Get() to save CPU
+      if (moptions_.memtable_whole_key_filtering) {
+        may_contain = bloom_filter_->MayContain(
```

Comment:
We'll still end up probing the bloom filter one by one for each key, which doesn't allow the cache access to be optimized. The range can be pushed down to the bloom filter implementation. Then all the relevant cachelines can be prefetched in one shot, followed by the probes here - https://github.com/facebook/rocksdb/blob/master/util/dynamic_bloom.h#L157. We do something similar in the full filter block - https://github.com/facebook/rocksdb/blob/master/table/block_based/full_filter_block.cc#L191-L218